### PR TITLE
py-numpy: add v2.0.0

### PR DESCRIPTION
**Pull request by adamjstewart**
https://github.com/numpy/numpy/releases/tag/v2.0.0

Excited to see how many packages break. We _could_ pin the default version to v1, but I'd rather fix any dependency bugs that pop up instead.

### CODE REVIEW

**Comment by adamjstewart** [comment]
Extremely helpful! https://github.com/numpy/numpy/issues/26191

**Comment by rgommers** [comment]
This seems pretty happy so far, modulo the triggering issues in some pipelines.

> Extremely helpful! [numpy/numpy#26191](https://github.com/numpy/numpy/issues/26191)

Glad it's useful:) For any incompatibilities we encounter here, I'd be happy to extend the compat table.

Most packages needed at least a few minor source-level changes. Most packages in that table also have ABI incompatibilities for wheels on PyPI - that's not going to matter for Spack, so it's very well possible that builds pass here even though they're marked as incompatible in [numpy/numpy#26191](https://github.com/numpy/numpy/issues/26191).

File: var/spack/repos/builtin/packages/treelite/package.py [hunk 1]
```diff
@@ -26,6 +26,8 @@ class Treelite(CMakePackage):
     depends_on("py-wheel", when="+python", type="build")
     depends_on("py-setuptools", when="+python", type="build")
     depends_on("py-numpy", when="+python", type=("build", "run"))
+    # https://github.com/dmlc/treelite/issues/560
+    depends_on("py-numpy@:1", when="@4.2.0+python", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/treelite/package.py]:
```suggestion
    depends_on("py-numpy@:1", when="@:4.2.0+python", type=("build", "run"))
```

**Comment by adamjstewart** [comment]
Any idea what's wrong with the remaining build errors (Linux, OneAPI)?

### [h5py](https://gitlab.spack.io/spack/spack/-/jobs/11630390)

```
ImportError: numpy/_core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so: undefined symbol: pow8_h
```

### [petsc4py](https://gitlab.spack.io/spack/spack/-/jobs/11630375)

```
  In file included from src/petsc4py/PETSc.c:1242:
  src/petsc4py/include/petsc4py/numpy.h:5:10: fatal error: 'numpy/arrayobject.h' file not found
      5 | #include <numpy/arrayobject.h>
        |          ^~~~~~~~~~~~~~~~~~~~~
  1 error generated.
```

### [scipy](https://gitlab.spack.io/spack/spack/-/jobs/11630386)

```
ImportError: numpy/_core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so: undefined symbol: pow8_h
```

**Comment by rgommers** [comment]
The `h5py` and `scipy` failures look the same, and unrelated to those packages. The `numpy` build with oneAPI seems broken. I can't locate it easily though - how do I navigate to the build of numpy itself?

**Comment by adamjstewart** [comment]
> I can't locate it easily though - how do I navigate to the build of numpy itself?

By default, Spack deletes the build directory (stage). If you use:
```console
spack install --keep-stage py-numpy@2.0.0 %oneapi
```
it won't delete the build directory. You can then run:
```console
spack cd py-numpy@2.0.0 %oneapi
```
to go to the build directory. Best source of documentation on this is probably https://spack-tutorial.readthedocs.io/en/latest/tutorial_packaging.html#debugging-package-builds.

**Comment by rgommers** [comment]
I meant in CI. There doesn't seem to be a good way to navigate from the failing jobs to the ones they depend on. It's the `py-numpy` build in CI that is doing something strange here I think. I'm away from my Linux dev machine this week, so can't try locally.

**Comment by adamjstewart** [comment]
> I meant in CI.

Ah sorry. I actually don't know how. @scottwittenburg @kwryankrattiger? We're trying to go from a failing build to its successful parent, but the successful parent is minimized because nothing changed in the last commit.

Workaround: most recent commit that rebuilt numpy: https://gitlab.spack.io/spack/spack/-/jobs/11627167

**Comment by rgommers** [comment]
> Workaround: most recent commit that rebuilt numpy: https://gitlab.spack.io/spack/spack/-/jobs/11627167

That log is full of errors, it shouldn't be passing. Right at the top of the log is this:
```
    from . import overrides
  File "/home/software/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeh/morepadding/linux-ubuntu22.04-x86_64_v3/oneapi-2024.0.0/py-numpy-2.0.0-tfvgeycazgvoco6wao4i5nk6cpmmylo2/lib/python3.11/site-packages/numpy/_core/overrides.py", line 8, in <module>
    from numpy._core._multiarray_umath import (
ImportError: /home/software/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeh/morepadding/linux-ubuntu22.04-x86_64_v3/oneapi-2024.0.0/py-numpy-2.0.0-tfvgeycazgvoco6wao4i5nk6cpmmylo2/lib/python3.11/site-packages/numpy/_core/_multiarray_umath.cpython-311-x86_64-linux-gnu.so: undefined symbol: pow8_h
```

And after that every single test is failing.

* * *

`pow8_h` is only used in the SVML sources shipped with NumPy (https://github.com/numpy/SVML/blob/32bf2a98420762a63ab418aaa0a7d6e17eb9627a/linux/avx512/svml_z0_pow_d_ha.s#L91). I've never seen an issue there - my first guess is a OneAPI bug in AVX512 support. OneAPI isn't tested in NumPy CI unfortunately, due to its ultra-clunky packaging.

**Comment by adamjstewart** [comment]
Okay. I'll mark it as incompatible for now and we can refine that version/compiler bounds if/when we learn more.

**Comment by kwryankrattiger** [comment]
I don't know if I understand the question. Just because something isn't rebuilt in that CI run, the binaries that it uses in downstream jobs will be whatever is in the build cache for that hash. Sometimes you have to look back in a PR to find the job that built that specific binary, or it was built previously on develop.

**Review by tldahlgren — DISMISSED** [review]
Confirmed the version sha256. Perused the rest and it LGTM.

**Comment by adamjstewart** [comment]
> ../meson.build:138:12: ERROR: Program 'pythran' not found or not executable

I know we had this problem before, but I don't remember how we solved it.

**Comment by rgommers** [comment]
> > ../meson.build:138:12: ERROR: Program 'pythran' not found or not executable
> 
> I know we had this problem before, but I don't remember how we solved it.

Hmm, this job doesn't have an outdated `setuptools` (like in https://github.com/spack/spack/pull/44967#issuecomment-2198529362), so maybe the problem is different. 

There were a few PRs that touched `py-pythran`, but nothing jumps out. Somehow entrypoint handling is messed up again. Here is a recent log that builds `py-pythran` 0.16.1: https://gitlab.spack.io/spack/spack/-/jobs/11404844. It only tests imports; the entrypoint's existence isn't checked in the tests run by Pythran.

**Comment by adamjstewart** [comment]
> It only tests imports; the entrypoint's existence isn't checked in the tests run by Pythran.

How would I test this on the command line? We can add a test to the `package.py` so that this is properly tested, although last I checked Python testing was broken (@tldahlgren)

Let's wait to see if my fix in #44967 magically helps for some unknown reason.

**Comment by rgommers** [comment]
> Let's wait to see if my fix in #44967 magically helps for some unknown reason.

Looks like it did:)

> How would I test this on the command line?

It's a matter of invoking `pythran` directly (e.g., `pythran --version`), rather than only importing it as a Python package.

**Review by alalazo — APPROVED** [review]

**Pull request merged**